modules/loader/source/relocations.cpp

```diff
@@ -645,88 +650,261 @@ bool resolveAArch64(const loader::Relocation &r, loader::ElfMap &map,
   }
   return true;
 }
+
+// Sets the immediate field of an I-type instruction [31:20]
+void writeITypeImm(uint32_t imm, uint8_t *insn_addr) {
+  uint32_t insn;
+  cargo::read_little_endian(&insn, insn_addr);
+  insn = setBitRange(insn, imm, 20, 12);
+  cargo::write_little_endian(insn, insn_addr);
+}
+
+// Sets the immediate field of an U-type instruction [31:12]
+void writeUTypeImm(uint32_t imm, uint8_t *insn_addr) {
+  uint32_t insn;
+  cargo::read_little_endian(&insn, insn_addr);
+  insn = setBitRange(insn, imm, 12, 20);
+  cargo::write_little_endian(insn, insn_addr);
+}
+
+uint32_t getLo12(uint64_t addr) {
+  return static_cast<uint32_t>(getBitRange(addr, 0, 12));
+}
+
+std::optional<uint32_t> getHi20(int64_t addr) {
+  auto signExtend32 = [](int64_t val) {
+    return static_cast<int64_t>(val << 32) >> 32;
+  };
+  addr += 0x800;
```

Comment:
what's this bit for?

Reply:
That's just how HI20 relocations are resolved. Think of it because you're taking it and adding a sign-extended LO12 value to it later.

If you had to resolve `0xDEADBEEF` and took the pieces apart naively:

HI20 = `0xDEADB` -> `0xDEADB000`
LO12 = `0xEEF` -> `0xFFFFFEEF`

If you added them up you'd get `0xDEADB000` + `0xFFFFFEEF` = `0xDEADAEEF` which isn't correct.

So instead you take:

HI20 = `0xDEADB` + `0x1` (skipping the addition of `0x800` then shifting right by 12) -> `0xDEADC000`
LO12 = `0xEEF` -> `0xFFFFFFEEF`

Adding those together gets the correct 32-bit address: `0xDEADC000` + `0xFFFFEEF` = `0xDEADBEEF`.

I didn't document it explicitly as it's just taken as a given of how you do it in the ELF documentation

https://github.com/riscv-non-isa/riscv-elf-psabi-doc/blob/master/riscv-elf.adoc#absolute-addresses
https://github.com/riscv-non-isa/riscv-elf-psabi-doc/blob/master/riscv-elf.adoc#pc-relative-symbol-addresses

Reply:
okay that makes sense, maybe just link to the documentation.

Reply:
done